### PR TITLE
[SR-7111] Class function override `T??` to `T`.

### PR DESCRIPTION
**Pull request by tarunon**
Resolves [SR-7111](https://bugs.swift.org/browse/SR-7111).

### CODE REVIEW

**Comment by slavapestov** [comment]
I’d rather we didn’t make this change. Ideally method override matching would be identical to the type checker’s subtype relationship. But I’m the absence of that, just adding new special cases is probably not s good idea.

**Comment by jrose-apple** [comment]
Override matching is already a different thing than subtypes, and I think the point is that this isn't a *new* special case; it's fixing the logic for an existing special case. FWIW this is already a reusable API, TypeBase::matches.

**Comment by slavapestov** [comment]
My point is exactly that the only reason override matching is different than subtyping is historical accident, and we should either leave it alone or combine the two.

**Comment by tarunon** [comment]
I think Differences is not historical accident, maybe it is about Existential.
```swift
class Foo {
    func get() -> Animal { fatalError() }
}

class Bar: Foo {
    override func get() -> Cat { fatalError() }
}
```
It seems be compile error today, can we solve it already?

**Comment by slavapestov** [comment]
@tarunon This example could be made to work pretty easily I think. SILGen can already emit the right thunk there because it supports all subtype conversions — what’s missing is the type checker admitting this case as an override. But my point is that it would be better in the long run to actually use the same code we use when solve subtype constraints, instead of using bespoke code to match overrides.

**Comment by slavapestov** [comment]
@rudkx might have some ideas, I think this might be related to his Type::join work.

**Comment by rudkx** [comment]
There is definitely non-accidental history behind why we cannot just test subtyping here.

We allow overriding Objective-C methods that take unaudited pointer arguments (imported as IUOs) with Swift methods that take non-optional arguments.

There may be other reasons as well.

File: test/decl/class/override.swift [hunk 1]
```diff
@@ -50,6 +86,18 @@ class B : A {
...
+  override func param_subclass_uoptional(_ x : A!) {}
+  func param_subclass_uoptional_rev(_ x : B) {}
+  override func param_subclass_optional_uoptional(_ x : A?) {}
+  func param_subclass_optional_uoptional_rev(_ x : B!) {}
+  override func param_nonclass_nested_optional(_ x : Int??) {}
+  func param_nonclass_nested_optional_rev(_ x : Int) {}
+  override func param_class_nested_optional(_ x : B??) {}
+  func param_class_nested_optional_rev(_ x : B) {}
+  override func param_subclass_nested_optional(_ x : A??) {}
+  func param_subclass_nested_optional_rev(_ x : B) {}
```
Comment (rudkx) [inline test/decl/class/override.swift]: Can you please add tests for things which should *not* be allowed as well?

**Comment by tarunon** [comment]
I know the various problems that optional covariances, I think T to T? subtyping should be abolished, but I also understand it is necessary for compatibility with Objective-C. Perhaps it is better to allow optional covariances in only `@ objc` attribute. I would like to discuss it in the swift-evolution.  
I tried to generalize Optional's subtyping according to the current subtyping rule in this PR. If we need a big modification (such as merging common subtyping and override), I think it would be better to discuss in swift-evolution.

**Comment by tarunon** [comment]
BTW, I try to using common subtyping way in this week end. @slavapestov

**Comment by tarunon** [comment]
I add property test cases, it should be failure case. Shall I add other test cases? @rudkx 
I try to use `TC::isSubtypeOf` instead of `Type::matches`, it seems walk well(also Existential case).

**Comment by rudkx** [comment]
@tarunon When you say that using `isSubtypeOf` works well, what tests did you run?

I would expect that to fail for the cases I already mentioned regarding the special-case for overriding unaudited Objective-C methods.

**Comment by tarunon** [comment]
I replace `matches` to `isSubtypeOf` in this line, and every test cases was passed.
https://github.com/apple/swift/blob/72e4224dc96ed107f2d136fc9952731794d08cfa/lib/Sema/TypeCheckDecl.cpp#L6402

But I found 1 more use `matches` at this line, https://github.com/apple/swift/blob/72e4224dc96ed107f2d136fc9952731794d08cfa/lib/Sema/TypeCheckDecl.cpp#L6155
I think I should add some of implements.

**Comment by tarunon** [comment]
I close PR cuz I can't see long time 😞
If I can, try make PR again 🧐

**Pull request closed without merging**